source/common/buffer/watermark_buffer.cc

```diff
@@ -147,35 +150,67 @@ void WatermarkBuffer::checkHighAndOverflowWatermarks() {
-    size_class_account_sets_[new_class].insert(account);
-  } else if (current_class >= 0 && new_class == -1) {
+    ASSERT(!size_class_account_sets_[new_class.value()].contains(account));
+    size_class_account_sets_[new_class.value()].insert(account);
+  } else if (current_class >= 0u && !new_class.has_value()) {
     // No longer track
-    ASSERT(size_class_account_sets_[current_class].contains(account));
-    size_class_account_sets_[current_class].erase(account);
+    ASSERT(size_class_account_sets_[current_class.value()].contains(account));
+    size_class_account_sets_[current_class.value()].erase(account);
   } else {
     // Moving between buckets
-    ASSERT(size_class_account_sets_[current_class].contains(account));
-    ASSERT(!size_class_account_sets_[new_class].contains(account));
-    size_class_account_sets_[new_class].insert(
-        std::move(size_class_account_sets_[current_class].extract(account).value()));
+    ASSERT(size_class_account_sets_[current_class.value()].contains(account));
+    ASSERT(!size_class_account_sets_[new_class.value()].contains(account));
+    size_class_account_sets_[new_class.value()].insert(
+        std::move(size_class_account_sets_[current_class.value()].extract(account).value()));
   }
 }
 
 void WatermarkBufferFactory::unregisterAccount(const BufferMemoryAccountSharedPtr& account,
-                                               int current_class) {
-  if (current_class >= 0) {
-    ASSERT(size_class_account_sets_[current_class].contains(account));
-    size_class_account_sets_[current_class].erase(account);
+                                               absl::optional<uint32_t> current_class) {
+  if (current_class.has_value()) {
+    ASSERT(size_class_account_sets_[current_class.value()].contains(account));
+    size_class_account_sets_[current_class.value()].erase(account);
+  }
+}
+
+void WatermarkBufferFactory::resetAccountsGivenPressure(float pressure) {
+  ASSERT(pressure >= 0.0 && pressure <= 1.0, "Provided pressure is out of range [0, 1].");
+
+  // Compute buckets to clear
+  const uint32_t buckets_to_clear = std::min<uint32_t>(
+      std::floor(pressure * BufferMemoryAccountImpl::NUM_MEMORY_CLASSES_) + 1, 8);
+  uint32_t bucket_idx = BufferMemoryAccountImpl::NUM_MEMORY_CLASSES_ - buckets_to_clear;
+
+  ENVOY_LOG_MISC(warn, "resetting streams in buckets >= {}", bucket_idx);
+
+  // Clear buckets
+  while (bucket_idx < BufferMemoryAccountImpl::NUM_MEMORY_CLASSES_) {
+    ENVOY_LOG_MISC(warn, "resetting {} streams in bucket {}.",
+                   size_class_account_sets_[bucket_idx].size(), bucket_idx);
+
+    auto it = size_class_account_sets_[bucket_idx].begin();
+    while (it != size_class_account_sets_[bucket_idx].end()) {
+      auto next = std::next(it);
+      // This will trigger an erase, which avoids rehashing and invalidates the
+      // iterator *it*. *next* is still valid.
+      (*it)->resetDownstream();
```

Comment:
outside the scope of this PR but I'd love a rename from this to something along the lines of tooMuchMemoryUsed() which we then used to reset the downstream.

IMO buffers shouldn't be thinking about upstream and downstream - if for instance a health checker was using too much memory that should reset the upstream connection since there is no downstream :-)